### PR TITLE
Embeds

### PR DESCRIPTION
**Pull request by mgerring**

### CODE REVIEW

File: assets/js/OutputIframe.js [hunk 1]
```diff
@@ -1,28 +1,120 @@
...
+    }
+  }
+
+  setPublic() {
+    store.dispatch(setWorkflowPublicAction(this.props.workflow.id, true));
+  }
+
+  renderSetPublicModal() {
+    return (
+      <Modal isOpen={this.state.setPublicModalOpen} toggle={this.toggleModals} className='test-setpublic-modal'>
```
Comment (jstray) [inline assets/js/OutputIframe.js]: Is this the same code as the public modal in (IIRC) WorkflowMetadata? Should be a shared component if so.

File: assets/js/OutputIframe.js [hunk 2]
```diff
@@ -1,28 +1,120 @@
...
+        {this.renderSetPublicModal()}
+        {this.renderEmbedCodeModal()}
+        <div className="outputpane-iframe-control-overlay">
+          <div className="btn icon icon-code" onClick={this.toggleModal} />
+        </div>
+        <iframe ref={this.setRef} src={this.state.src} />
+      </div>
+    );
   }
 }
```
Comment (jstray) [inline assets/js/OutputIframe.js]: Needs PropTypes

File: assets/js/workflow-reducer.js [hunk 3]
```diff
@@ -122,6 +123,26 @@ registerReducerFunc(RELOAD_WORKFLOW + '_FULFILLED', (state, action) => {
...
+  return function (dispatch) {
+    return (
+
+      dispatch({
+        type: SET_WORKFLOW_PUBLIC,
+        payload: api.setWorkflowPublic(workflowId, isPublic)
+      }).then( () => {
+
+      dispatch(
+        reloadWorkflowAction()
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: Why is the reload necessary? Isn't the public flag in the redux store, and therefore ultimately passed as a prop to the navbar where it's displayed?

File: server/serializers.py [hunk 4]
```diff
@@ -130,14 +130,14 @@ class WorkflowSerializer(serializers.ModelSerializer):
     owner_name = serializers.SerializerMethodField()
 
     def get_read_only(self, obj):
-        return obj.read_only(self.context['user'])
+        return obj.read_only(self.context.get('user', False))
```
Comment (jstray) [inline server/serializers.py]: Why is the False required? Could use a comment

File: server/views/embed.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+from django.template.response import TemplateResponse
+from rest_framework.renderers import JSONRenderer
+from server.models import WfModule
+from server.serializers import WorkflowSerializer, WfModuleSerializer
+
+def embed(request, wfmodule_id):
+    try:
```
Comment (jstray) [inline server/views/embed.py]: this can be shortened to get_object_or_404

File: templates/base.html [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {% load analytics_tags %}
 
 <!DOCTYPE html>
-<html lang="en">
+<html lang="en" {% block html_tag_attrs %}{% endblock %}>
```
Comment (jstray) [inline templates/base.html]: Why is this needed?

File: assets/js/workflow-reducer.js (hunk 3 above)
Comment (mgerring) [reply assets/js/workflow-reducer.js]: It may not be necessary, but we have to deal with both 'public' and 'isReadOnly', and the way that both 'public' and 'readOnly' properties are handled needs a full re-write that I wasn't ready to deal with here.

File: server/serializers.py (hunk 4 above)
Comment (mgerring) [reply server/serializers.py]: In case we have a request with no user, will add a comment

File: server/views/embed.py (hunk 5 above)
Comment (mgerring) [reply server/views/embed.py]: No it can't for two reasons: 1) Django doesn't support per-view 404 pages using get_object_or_404, if you want a custom 404 you have to do it manually. Also, this doesn't send a 404. It sets wf_module and workflow to 'none' and renders the same template, so we can handle both botched embed codes and embed codes that used to work but have been made private.

File: templates/base.html (hunk 6 above)
Comment (mgerring) [reply templates/base.html]: Can take this out now

File: assets/js/OutputIframe.js (hunk 1 above)
Comment (mgerring) [reply assets/js/OutputIframe.js]: Sort of, but that works because WorkflowMetadata and NavBar share state. The entire handling of public and readOnly needs refactoring using redux. Then we can make this shared code.

File: templates/base.html (hunk 6 above)
Comment (mgerring) [reply templates/base.html]: I take that back, this is necessary to get the document height to 100% to make sure the iframe is sized correctly.
